kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetStatsReader.java

```diff
@@ -0,0 +1,238 @@
+            int scale = ((DecimalLogicalTypeAnnotation) logicalType).getScale();
+
+            DecimalType decimalType = (DecimalType) dataType;
+
+            // Check the scale is same in both the Delta data type and the Parquet Logical Type
+            checkArgument(
+                    scale == decimalType.getScale(),
+                    "DECIMAL column has different scale compared to the input data: %s", scale);
+
+            // Decimal is stored either as int, long or binary. Decode the stats accordingly.
+            BigDecimal decimalStatValue;
+            if (statistics instanceof IntStatistics) {
+                decimalStatValue = BigDecimal.valueOf((Integer) statValue).movePointLeft(scale);
+            } else if (statistics instanceof LongStatistics) {
+                decimalStatValue = BigDecimal.valueOf((Long) statValue).movePointLeft(scale);
+            } else if (statistics instanceof BinaryStatistics) {
+                BigInteger base = new BigInteger(getBinaryStat(statistics, decodeMin));
+                decimalStatValue = new BigDecimal(base, scale);
+            } else {
+                throw new UnsupportedOperationException(
+                        "Unsupported stats type for Decimal: " + statistics.getClass());
+            }
+            return Literal.ofDecimal(decimalStatValue, decimalType.getPrecision(),
+                    decimalType.getScale());
+        } else if (dataType instanceof DateType) {
+            checkArgument(
+                    statistics instanceof IntStatistics,
+                    "Column with DATE type contained invalid statistics: %s", statistics);
+            return Literal.ofDate((Integer) statValue); // stats are stored as epoch days
+        } else if (dataType instanceof StringType) {
+            byte[] binaryStat = getBinaryStat(statistics, decodeMin);
+            return Literal.ofString(new String(binaryStat, UTF_8));
+        } else if (dataType instanceof BinaryType) {
+            return Literal.ofBinary(getBinaryStat(statistics, decodeMin));
+        }
+
+        throw new IllegalArgumentException("Unsupported stats data type: " + statValue);
+    }
+
+    private static Optional<Statistics<?>> mergeMetadataList(
+            Collection<ColumnChunkMetaData> metadataList) {
+        if (hasInvalidStatistics(metadataList)) {
+            return Optional.empty();
+        }
+
+        return metadataList.stream()
+                .<Statistics<?>>map(ColumnChunkMetaData::getStatistics)
+                .reduce((statsA, statsB) -> {
+                    statsA.mergeStatistics(statsB);
+                    return statsA;
+                });
+    }
+
+    private static boolean hasInvalidStatistics(Collection<ColumnChunkMetaData> metadataList) {
+        // If any row group does not have stats collected, stats for the file will not be valid
+        return metadataList.stream().anyMatch(metadata ->
```

Comment:
same here. i wonder if we should avoid `stream` on the hot path. can you confirm: is this on any sort of hot path? seems like it could be happening multiple times for every file?

Reply:
This is not the hotpath. It is executed once per file.